[OMPIRBuilder] Support runtime number of teams and threads, and SPMD mode

skatrak opened the pull request:
This patch introduces a `TargetKernelRuntimeAttrs` structure to hold host-evaluated `num_teams`, `thread_limit`, `num_threads` and trip count values passed to the runtime kernel offloading call.

Additionally, kernel type information is used to influence target device code generation and the `IsSPMD` flag is replaced by `ExecFlags`, which provide more granularity.

skatrak commented:
PR stack:
- #116048
- #116049
- #116050
- #116051
- #116052

jdoerfert commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -6758,6 +6793,27 @@ static Expected<Function *> createOutlinedFunction(
+    Func->addFnAttr(TargetFeaturesAttr);
+
+  if (OMPBuilder.Config.isTargetDevice()) {
+    std::vector<llvm::WeakTrackingVH> LLVMCompilerUsed;
```
Why do you need a vector here, especially of tracking VH?
I guess what you want is:
```
auto *V = emitExecMode(...)
emitUsed("llvm.compiler.used", {V}, OMPBuilder.M);
```

skatrak replied:
Done.

jdoerfert commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -7427,13 +7521,17 @@ emitTargetCall(OpenMPIRBuilder &OMPBuilder, IRBuilderBase &Builder,
     OpenMPIRBuilder::TargetGenArgAccessorsCallbackTy ArgAccessorFuncCB,
     SmallVector<DependData> Dependencies, bool HasNowait) {
+  assert((!RuntimeAttrs.LoopTripCount || IsSPMD) &&
+         "trip count not expected if IsSPMD=false");
```
This might be true right now, but it should not be true.

skatrak replied:
Removed.

jdoerfert commented on llvm/unittests/Frontend/OpenMPIRBuilderTest.cpp:
```diff
@@ -6122,8 +6122,10 @@ TEST_F(OpenMPIRBuilderTest, TargetRegion) {
   OMPBuilder.setConfig(Config);
   F->setName("func");
+  F->addFnAttr("target-cpu", "x86-64");
+  F->addFnAttr("target-features", "+mmx,+sse");
```
unrelated to this PR

jdoerfert commented on llvm/unittests/Frontend/OpenMPIRBuilderTest.cpp:
```diff
@@ -6378,6 +6439,204 @@ TEST_F(OpenMPIRBuilderTest, TargetRegionDevice) {
+  auto SimpleArgAccessorCB =
+      [&](llvm::Argument &, llvm::Value *, llvm::Value *&,
+          llvm::OpenMPIRBuilder::InsertPointTy,
+          llvm::OpenMPIRBuilder::InsertPointTy CodeGenIP) {
```
all the places in llvm/, no `llvm::`, it's not MLIR.

skatrak replied:
Done.

skatrak committed:
[OMPIRBuilder] Support runtime number of teams and threads, and SPMD mode

This patch introduces a `TargetKernelRuntimeAttrs` structure to hold
host-evaluated `num_teams`, `thread_limit`, `num_threads` and trip count values
passed to the runtime kernel offloading call.

Additionally, kernel type information is used to influence target device code
generation and the `IsSPMD` flag is replaced by `ExecFlags`, which provide more
granularity.

skatrak commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -6727,8 +6727,43 @@ FunctionCallee OpenMPIRBuilder::createDispatchDeinitFunction() {
+      Twine(FunctionName, "_exec_mode"));
+  GVMode->setVisibility(llvm::GlobalVariable::ProtectedVisibility);
+  LLVMCompilerUsed.emplace_back(GVMode);
+}
```
Moved to OMPIRBuilder and simplified implementation.

skatrak commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -6758,6 +6793,27 @@ static Expected<Function *> createOutlinedFunction(
+
+  auto TargetFeaturesAttr = ParentFn->getFnAttribute("target-features");
+  if (TargetFeaturesAttr.isStringAttribute())
+    Func->addFnAttr(TargetFeaturesAttr);
```
Good catch, I moved that to another PR right after this one in the stack.

skatrak reviewed (commented):
Thank you @jdoerfert for the review. Your comments should be addressed now.

DominikAdamski reviewed (approved):
LGTM

skatrak commented:
The PR stack should be almost ready to be merged in the next few days (buildbot failures are unrelated). If there are any remaining blockers from your side, let me know @jdoerfert.
